staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/vault.go

```diff
@@ -0,0 +1,211 @@
+}
+
+func checkAuthConfig(config *EnvelopeConfig) error {
+	var count uint
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return fmt.Errorf("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return fmt.Errorf("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return fmt.Errorf("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return fmt.Errorf("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, fmt.Errorf("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	cipher := strings.Replace(data, key, vaultPrefix, 1)
+
+	//plain, _, err := s.client.decrypt(key, cipher)
+	plain, err := s.withRefreshToken((*clientWrapper).decrypt, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken((*clientWrapper).encrypt, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
+	// "vault:" is unnecessary, replace it with key name.
+	return strings.Replace(cipher, vaultPrefix, key, 1), nil
+}
+
+// The function type for clientWrapper.encrypt and clientWrapper.decrypt.
+type encryptOrDecryptFunc func(*clientWrapper, string, string) (string, error)
+
+func (s *vaultEnvelopeService) withRefreshToken(f encryptOrDecryptFunc, key, data string) (string, error) {
+	// Execute operation first time.
+	s.rwmutex.RLock()
+	result, err := f(s.client, key, data)
+	s.rwmutex.RUnlock()
+	if err == nil || s.config.Token != "" {
+		return result, err
+	}
+	_, ok := err.(*forbiddenError)
+	if !ok {
+		return result, err
+	}
+
+	// The request is forbidden, refresh token and execute operation again.
+	s.rwmutex.Lock()
+	defer s.rwmutex.Unlock()
+	err = s.client.refreshToken(s.config)
+	if err != nil {
+		return result, err
+	}
+
+	secret, requestErr := f(s.client, key, data)
```

Comment:
To me this still appears like it can fail due to race conditions if the token has a limited number of uses and/or there are enough goroutines making calls to encrypt/decrypt. 

Have you ran this simulating many go routines to the single client?

Reply:
Oh, i see what you changed here. This shouldn't technically error anymore; however, it's going to have potentially bad performance when multiple goroutines all fail during the read lock calls and block on the write mutex. I think in its current state it's fine, but I feel there may be push back upstream.

Reply:
Given the usage, the contention is less likely:
1. There is caching of DEK in the framework code, so most decrypt operations will not require call to KMS providers.
2. Encrypt operations will not be frequent as encrypt will happen only during creation or update of a secret.
3. It is not expected that multiple secrets will be created or updated concurrently frequently.
4. If token policy is chosen to have a reasonable num of use/expiry, 403 will not be that frequent.

As long as technically it is guaranteed to work in rare worst case scenario at a degraded performance, I think it should be good. I can put the explanation in comments so that someone reads and validates the assumptions.

Reply:
👍 great explanation. Thanks!